### PR TITLE
[YTA-1637] Initial integration with card-payment-history. With @nigel-perkins.

### PR DESCRIPTION
**Pull request by erasmos**

### CODE REVIEW

File: app/config/EnvironmentConfiguration.scala [hunk 1]
```diff
@@ -61,7 +62,8 @@ object EnvironmentConfiguration {
   }
 
   lazy val emisStatusMetricsPublisher = EmisStatusMetricsPublisher(merchants)
-  lazy val emisReportTransformer = EmisReportTransformer(merchants)
+  lazy val cardPaymentHistoryConnector = CardPaymentHistoryConnector
```
Comment (anshulbajpai) [inline app/config/EnvironmentConfiguration.scala]: Any reason that these dependencies are lazy?

File: app/transformer/EmisReportTransformer.scala [hunk 2]
```diff
@@ -16,80 +16,110 @@
...
+
+    val validMerchantTransactions: Iterator[(MerchantTransaction, String)] = getValidMerchantTransactions(emisReport)
+
+    val iteratorOfRcsPaymentFutures: Iterator[Future[RcsPayment]] = validMerchantTransactions.zipWithIndex.map { case ((merchantTransaction, taxRef), index) =>
+      buildPayment(merchantTransaction, index + 1, merchantTransaction.merchant.merchantId, taxRef)
+    }
+
+    val iteratorOfRcsPaymentsFuture: Future[Iterator[RcsPayment]] = Future.sequence(iteratorOfRcsPaymentFutures)
+
+    iteratorOfRcsPaymentsFuture
```
Comment (anshulbajpai) [inline app/transformer/EmisReportTransformer.scala]: Inline iteratorOfRcsPaymentsFuture

File: app/utils/CommissionUtils.scala [hunk 3]
```diff
@@ -18,11 +18,9 @@ package utils
...
 
-object Commission {
-
-  val rate = 1.5
-
-  def apply(total: Pence): Pence = calculate(total, rate)
-  private def calculate(total: Pence, rate: Double) = total map (t => math.max(1,  math.round(t / (100 + rate) * rate)))
+object CommissionUtils {
+  
+  def calculate(grossAmountInPence: Pence, rateAsPercentage: BigDecimal): Pence = calculate(grossAmountInPence, rateAsPercentage.toDouble)
```
Comment (anshulbajpai) [inline app/utils/CommissionUtils.scala]: Calculate What?

File: test/unit/connectors/CardPaymentHistoryConnectorSpec.scala [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+import org.mockito.Matchers
+import org.mockito.Mockito._
+import org.scalatest.concurrent.ScalaFutures
+import org.scalatest.mock.MockitoSugar
+import uk.gov.hmrc.play.http.{BadRequestException, HeaderCarrier, HttpGet}
+import uk.gov.hmrc.play.test.UnitSpec
+
+import scala.concurrent.{Await, Future}
+
+class CardPaymentHistoryConnectorSpec extends UnitSpec  with MockitoSugar {
```
Comment (anshulbajpai) [inline test/unit/connectors/CardPaymentHistoryConnectorSpec.scala]: I don't see assertions around returned value from this method.

File: app/transformer/EmisReportTransformer.scala [hunk 5]
```diff
@@ -16,80 +16,110 @@
...
-      .filter { merchantTransaction =>
+  private def toRcsPayments(emisReport: Iterator[MerchantTransaction])(implicit headerCarrier: HeaderCarrier): Future[Iterator[RcsPayment]] = {
+
+    val validMerchantTransactions: Iterator[(MerchantTransaction, String)] = getValidMerchantTransactions(emisReport)
+
+    val iteratorOfRcsPaymentFutures: Iterator[Future[RcsPayment]] = validMerchantTransactions.zipWithIndex.map { case ((merchantTransaction, taxRef), index) =>
+      buildPayment(merchantTransaction, index + 1, merchantTransaction.merchant.merchantId, taxRef)
+    }
+
+    val iteratorOfRcsPaymentsFuture: Future[Iterator[RcsPayment]] = Future.sequence(iteratorOfRcsPaymentFutures)
```
Comment (anshulbajpai) [inline app/transformer/EmisReportTransformer.scala]: I hope this is sending requests concurrently

File: app/config/EnvironmentConfiguration.scala (hunk 1 above)
Comment (erasmos) [inline app/config/EnvironmentConfiguration.scala]: Not sure; it was like that already.

File: app/transformer/EmisReportTransformer.scala (hunk 2 above)
Comment (erasmos) [inline app/transformer/EmisReportTransformer.scala]: We wanted to make it clearer, what Future.sequence was doing, but if you wish.

File: app/config/EnvironmentConfiguration.scala (hunk 1 above)
Comment (anshulbajpai) [inline app/config/EnvironmentConfiguration.scala]: cardPaymentHistoryConnector is introduced new. Can we atleast have cardPaymentHistoryConnector non lazy if it is possible?

File: app/transformer/EmisReportTransformer.scala (hunk 2 above)
Comment (anshulbajpai) [inline app/transformer/EmisReportTransformer.scala]: Its fine then.

File: app/config/EnvironmentConfiguration.scala (hunk 1 above)
Comment (erasmos) [inline app/config/EnvironmentConfiguration.scala]: We just tried removing the lazy, and many tests failed, so leaving it.

File: app/transformer/EmisReportTransformer.scala (hunk 2 above)
Comment (erasmos) [inline app/transformer/EmisReportTransformer.scala]: Inlined.

File: app/utils/CommissionUtils.scala (hunk 3 above)
Comment (erasmos) [inline app/utils/CommissionUtils.scala]: Given that it's in the CommissionUtils object, we thought that calculate would refer to a commission.

File: app/config/EnvironmentConfiguration.scala (hunk 1 above)
Comment (anshulbajpai) [inline app/config/EnvironmentConfiguration.scala]: Ok. Thanks

File: test/unit/connectors/CardPaymentHistoryConnectorSpec.scala (hunk 4 above)
Comment (erasmos) [inline test/unit/connectors/CardPaymentHistoryConnectorSpec.scala]: We don't care about the return value, as those scenarios are asserted the metric calls.

The return value is already covered above, in the Specs.

File: app/transformer/EmisReportTransformer.scala (hunk 5 above)
Comment (erasmos) [inline app/transformer/EmisReportTransformer.scala]: Yes it is.

File: test/unit/connectors/CardPaymentHistoryConnectorSpec.scala (hunk 4 above)
Comment (anshulbajpai) [inline test/unit/connectors/CardPaymentHistoryConnectorSpec.scala]: I could see only test related to metrics in this specs file. I don't see any assertions around the returned value.

**Comment by AndyNewton** [comment]
Merging, as I have credentials; reviewed by @anshulbajpai.

**Pull request merged**